tools/eslint-plugin-cdk/lib/rules/invalid-cfn-imports.ts

```diff
@@ -0,0 +1,151 @@
+          node: node,
+        });
+      }
+    },
+
+    // This captures using `xxx.CfnConstruct` as an identifier
+    Identifier: node => {
+      const typeAnnotation = (node as any).typeAnnotation?.typeAnnotation;
+      const type = typeAnnotation?.typeName;
+      if (type?.type === 'TSQualifiedName') {
+        const result = checkLeftAndRightForCfn(type);
+        if (result) {
+          reportErrorIfImportedLocationIsNotValid(context, node, result.name, result.location);
+        }
+      }
+      if (node.name.startsWith('Cfn') && node.parent.type === 'MemberExpression' && node.parent.object.type === 'Identifier') {
+        // new xxx.CfnConstruct();
+        reportErrorIfImportedLocationIsNotValid(context, node, node.name, node.parent.object.name);
+      }
+    },
+  };
+}
+
+function reportErrorIfImportedLocationIsNotValid(context: Rule.RuleContext, node: Identifier, name: string, location: string) {
+  const moduleSpecifier = namespaceImports[location];
+  if (moduleSpecifier.endsWith('generated')) {
+    return;
+  }
+  if (moduleSpecifier.startsWith('.')) {
+    // import * as name from './some/relative/path/not/ending/in/generated'; name.CfnConstruct();
+    context.report({
+      message: 'To allow rewriting imports when generating v2 experimental modules, `' + name + '` must be imported by name from its specific .generated location.',
+      node: node,
+    });
+  } else if (checkIfImportedLocationIsAnAlphaPackage(moduleSpecifier)) {
+    // import * as name from '@aws-cdk/another-alpha-module'; name.CfnConstruct();
+    context.report({
+      message: 'To allow rewriting imports when generating v2 experimental modules, `' + name + '` must be imported by name and separate from non-L1 imports, since it is being imported from an experimental package: ' + moduleSpecifier,
+      node: node,
+    });
+  }
+}
+
+function checkIfImportedLocationIsAnAlphaPackage(_moduleName: string): boolean {
+  // TODO
+  return true;
+}
+
+function inAlphaPackage(filename: string): boolean {
+  const filenameSplit = filename.split('/');
+  const awsCdkNamespaceIndex = filenameSplit.findIndex(e => e.match('@aws-cdk'));
+  if (awsCdkNamespaceIndex !== -1) {
+    const moduleDir = filenameSplit.slice(0, awsCdkNamespaceIndex + 2).join('/');
+    const pkg = JSON.parse(fs.readFileSync(path.join(moduleDir, 'package.json'), { encoding: 'utf-8' }));
+
+    const separateModule = pkg['separate-module'];
+    if (separateModule !== undefined) {
+      return separateModule;
+    }
+
+    const maturity = pkg.maturity;
+    if (maturity !== 'experimental' && maturity !== 'developer-preview') {
+      return false;
+    }
+    // we're only interested in '@aws-cdk/' packages,
+    // and those that are JSII-enabled (so no @aws-cdk/assert)
+    return pkg.name.startsWith('@aws-cdk/') && !!pkg['jsii'];
+  }
+  return false;
+}
+
+function checkLeftAndRightForCfn(node: any): { name: string, location: string } | undefined {
```

Comment:
It doesn't look like the return type of this method is being used/checked, except for truthiness. Can this be simplified?

Reply:
It is used to print out the imported type and the location in the error message.